From 224629e9cb72238cec5c8290b3fcdee273f3b9cc Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sun, 7 Sep 2014 22:04:36 -0400
Subject: [iommu] core: Make iommu_group_get_for_dev() more robust

Message-id: <20140907220436.7201.14295.stgit@gir.stowe>
Patchwork-id: 91627
O-Subject: [RHEL7.1 PATCH 18/17] iommu/core: Make iommu_group_get_for_dev() more robust
Bugzilla: 1136537
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1136537
Upstream Status: This patch is in Joerg's iommu/fixes and iommu/next
branches.
  https://git.kernel.org/cgit/linux/kernel/git/joro/iommu.git/commit/?h=iommu/fixes&id=c4a783b89ee3fc1201510ecf204278da4ccb0993
  https://git.kernel.org/cgit/linux/kernel/git/joro/iommu.git/commit/?h=next&id=c4a783b89ee3fc1201510ecf204278da4ccb0993
As such, we expect it will be submitted to Linus in a v3.17-rcX
update.  Its an obvious fix for an oversight introduced in commit 104a1c1
- PATCH 12/17 of this series.
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7926508

commit c4a783b89ee3fc1201510ecf204278da4ccb0993
Author: Joerg Roedel <jroedel@suse.de>
Date:

iommu/core: Make iommu_group_get_for_dev() more robust

When a non-PCI device is passed to that function it might
pass group == NULL to iommu_group_add_device() which then
dereferences it and cause a crash this way. Fix it by
just returning an error for non-PCI devices.

Fixes: 104a1c13ac66e40cf8c6ae74d76ff14ff24b9b01
Cc: Alex Williamson <alex.williamson@redhat.com>
Acked-by: Alex Williamson <alex.williamson@redhat.com>
Signed-off-by: Joerg Roedel <jroedel@suse.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/iommu.c b/drivers/iommu/iommu.c
index d061c86..fde1964 100644
--- a/drivers/iommu/iommu.c
+++ b/drivers/iommu/iommu.c
@@ -674,15 +674,17 @@ static struct iommu_group *iommu_group_get_for_pci_dev(struct pci_dev *pdev)
  */
 struct iommu_group *iommu_group_get_for_dev(struct device *dev)
 {
- struct iommu_group *group = ERR_PTR(-EIO);
+ struct iommu_group *group;
  int ret;
 
  group = iommu_group_get(dev);
  if (group)
   return group;
 
- if (dev_is_pci(dev))
-  group = iommu_group_get_for_pci_dev(to_pci_dev(dev));
+ if (!dev_is_pci(dev))
+  return ERR_PTR(-EINVAL);
+
+ group = iommu_group_get_for_pci_dev(to_pci_dev(dev));
 
  if (IS_ERR(group))
   return group;
-- 
1.7.1